File: src/15-challenge/solution.py
def fn(data: list[dict[str, int | str]], sortBy: str) -> str:
    _data = sorted(
        data,
        key=lambda element: element[sortBy],
    )

    table: list[list[int | str]] = []
    header_lengths = list((0 for _ in enumerate(data[0])))

    for element in data:
        header_index = 0
        row: list[int | str] = []

        for key in element:
            value = element[key]
            header_lengths[header_index] = max(
                len(f"{value}"), header_lengths[header_index]
            )

            row.append(value)
            header_index = len(row)

        table.append(row)

    sep = "+"
    for col_length in header_lengths:
        sep += "+".rjust(col_length + 3, "-")

    header = "|"
    for i in range(len(table[0])):
        header += f" {chr(i + 65)}".ljust(header_lengths[i] + 1, " ") + " |"

    content = ""
    for row_data in table:
        for i, col_data in enumerate(row_data):
            content += f"| {col_data}".ljust(header_lengths[i] + 3, " ")

        content += "|\n"

    text_table = f"{sep}\n{header}\n{sep}\n{content}{sep}"

    return text_table

File: src/15-challenge/solution.js
/**
 * @param {Array<Object>} data - The data to draw the table
 * @param {string} sortBy - The field to sort the table
 * @returns {string}
 */
export default function fn(data, sortBy) {
	data.sort(
		typeof data[0][sortBy] === 'number' ? (a, b) => a[sortBy] - b[sortBy] : (a, b) => a[sortBy].localeCompare(b[sortBy])
	)

	/** @type {(number | string)[][]} */
	const table = []

	/** @type {number[]} */
	const headerLengths = Array.from({ length: Object.keys(data[0]).length }, () => 0)

	for (const element of data) {
		let headerIndex = 0

		/** @type {(number | string)[]} */
		const row = []

		for (const key in element) {
			const value = element[key]

			headerLengths[headerIndex] = Math.max(value.toString().length, headerLengths[headerIndex])

			row.push(value)
			headerIndex = row.length
		}

		table.push(row)
	}

	let sep = '+'
	for (const colLength of headerLengths) {
		sep += '+'.padStart(colLength + 3, '-')
	}

	let header = '|'
	for (let i = 0; i < table[0].length; i++) {
		header += ` ${String.fromCharCode(i + 65)}`.padEnd(headerLengths[i] + 1, ' ') + ' |'
	}

	let content = ''
	for (const rowData of table) {
		for (let i = 0; i < rowData.length; i++) {
			content += `| ${rowData[i]}`.padEnd(headerLengths[i] + 3, ' ')
		}

		content += '|\n'
	}

	const textTable = `${sep}\n${header}\n${sep}\n${content}${sep}`

	return textTable
}
